lib/srv/app/aws/aws.go

```diff
@@ -0,0 +1,146 @@
+	// Date value is specified using YYYYMMDD format.
+	Date string
+	// Region is an AWS Region.
+	Region string
+	// Service is an AWS Service.
+	Service string
+	// SignedHeaders is a  list of request headers that you used to compute Signature.
+	SignedHeaders []string
+	// The 256-bit Signature of the request.
+	Signature string
+}
+
+// ParseSigV4 AWS SigV4 credentials string sections.
+// AWS SigV4 header example:
+// Authorization: AWS4-HMAC-SHA256
+// Credential=AKIAIOSFODNN7EXAMPLE/20130524/us-east-1/s3/aws4_request,
+// SignedHeaders=host;range;x-amz-date,
+// Signature=fe5f80f77d5fa3beca038a248ff027d0445342fe2855ddc963176630326f1024
+func ParseSigV4(header string) (*SigV4, error) {
+	if header == "" {
+		return nil, trace.BadParameter("empty header")
+	}
+	sectionParts := strings.Split(header, " ")
+
+	m := make(map[string]string)
+	for _, v := range sectionParts {
+		kv := strings.Split(v, "=")
+		if len(kv) != 2 {
+			continue
+		}
+		m[kv[0]] = strings.TrimSuffix(kv[1], ",")
+	}
+
+	authParts := strings.Split(m[credentialAuthHeaderElem], "/")
+	if len(authParts) != 5 {
+		return nil, trace.BadParameter("invalid size of %q section", credentialAuthHeaderElem)
+	}
+
+	signature := m[signatureAuthHeaderElem]
+	if signature == "" {
+		return nil, trace.BadParameter("invalid signature")
+	}
+	var signedHeaders []string
+	if v := m[signedHeaderAuthHeaderElem]; v != "" {
+		signedHeaders = strings.Split(v, ";")
+	}
+
+	return &SigV4{
+		KeyID:     authParts[0],
+		Date:      authParts[1],
+		Region:    authParts[2],
+		Service:   authParts[3],
+		Signature: signature,
+		// Split semicolon-separated list of signed headers string.
+		// https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
+		// https://github.com/aws/aws-sdk-go/blob/main/aws/signer/v4/v4.go#L631
+		SignedHeaders: signedHeaders,
+	}, nil
+}
+
+// IsSignedByAWSSigV4 checks is the request was signed by AWS Signature Version 4 algorithm.
+// https://docs.aws.amazon.com/general/latest/gr/signing_aws_api_requests.html
+func IsSignedByAWSSigV4(r *http.Request) bool {
+	return strings.HasPrefix(r.Header.Get(authorizationHeader), AmazonSigV4AuthorizationPrefix)
+}
+
+// GetAndReplaceReqBody returns the request and replace the drained body reader with io.NopCloser
+// allowing for further body processing by http transport.
+func GetAndReplaceReqBody(req *http.Request) ([]byte, error) {
+	if req.Body == nil || req.Body == http.NoBody {
+		return []byte{}, nil
+	}
+	// req.Body is closed during drainBody call.
+	payload, err := drainBody(req.Body)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Replace the drained body with io.NopCloser reader allowing for further request processing by HTTP transport.
+	req.Body = io.NopCloser(bytes.NewReader(payload))
+	return payload, nil
+}
+
+// drainBody drains the body, close the reader and returns the read bytes.
+func drainBody(b io.ReadCloser) ([]byte, error) {
+	payload, err := io.ReadAll(b)
```

Comment:
I have check with 1GB file and for my local setup upload time looks like below:

Native `aws s3 cp`
```bash
time aws s3 cp ~/Downloads/1GB.bin s3://ice-berg-my-test-bucket/1G.bin 
1:34.13 total
```
tsh aws:
```bash
time tsh aws s3 cp ~/Downloads/1GB.bin s3://ice-berg-my-test-bucket/1G.bin
1:37.33 total
```

also the`tsh aws s3` upload process bar seems to be not as  fluent as in raw `aws s3 cp` command probably because larger flush intervals.